### DISCUSSION TITLE
Purpose of Over 100 subsequent flushes have occurred error...

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7916] [Usage Questions] [answered]
All:

I have some code that is giving me some unexepected behaviors recently. It's not at all clear yet, if the problem always existed, or if it is happening now in conjunction with some upgrades (to Python 3.9, more recent SQLA, etc).

I'm hitting the exception "Over 100 subsequent flushes have occurred..." thrown in _prepare_impl() in session.py

git blame on that line shows it's 8 years old!  So, this is probably not new SQLA behavior.

1) Would you consider changing the wording of this error? It leaves the impression that the flush()es in question are in client code. But what actually seems to be happening is that the session code itself tries up to 100 flush() calls, each time checking if the session._is_clean().

2) The "100" limit seems arbitrary. Is it ever ok for 88 flush()es to be necessary here? If so, why not 101? Why not throw after the first one? In other words, is it absolutely incorrect behavior in a program to "create new objects" or whatever it is that is actually causing _is_clean() to return false?

3) I'm using a custom session object that has some facilities to help in gevent/multi-processing SQLA usage. Possibly that's what's interfering with the _is_clean(). Can someone give me a high level idea of what the code in _prepare_impl(), which is attempting 100 flushes is intended to do?

Thanks,
A.

EDIT 1:  _is_clean() seems to return True is there are no modifications in the identity_map, no new and no deleted objects, I guess. If the flush() action happening in the _prepare_impl() ends up being a NOOP in my subclass of Session that I suppose would lead to this behavior, because then the internal flush doesn't affect the status of any dirty/etc objects...

... but I'd still like to understand 1-3 above.

EDIT 2: Looking at sqla's "main" branch I thought maybe the problem was due to (https://github.com/sqlalchemy/sqlalchemy/commit/cc46ea711df77540d5d658e9c7b3ab1e88288929) but looking at the installed SQLA code on my system, 1.4.35 does not seem to contain those changes. I think this problematic behavior appeared sometime between 1.4.31 and 1.4.35 of SQLA.

It's not clear to me how to navigate in github to the 1_4_31 code (can anyone tell me? Clicking the 1_4_31 tag doesn't take me to a full code tree, but only to the release notes???)

**Comment by zzzeek** [comment, marked as answer]
> All:
> 
> I have some code that is giving me some unexepected behaviors recently. It's not at all clear yet, if the problem always existed, or if it is happening now in conjunction with some upgrades (to Python 3.9, more recent SQLA, etc).
> 
> I'm hitting the exception "Over 100 subsequent flushes have occurred..." thrown in _prepare_impl() in session.py
> 
> git blame on that line shows it's 8 years old! So, this is probably not new SQLA behavior.
> 
>     1. Would you consider changing the wording of this error? It leaves the impression that the flush()es in question are in client code. 

that is exactly what this error indicates, that user code is how this condition is produced.

> But what actually seems to be happening is that the session code itself tries up to 100 flush() calls, each time checking if the session._is_clean().

it's not possible for the Session to do this on its own.  It only occurs when there are used-defined event handlers that are adding new items to the session within the flush process itself (or I suppose in theory, since you have said you have a custom Session subclass, if that subclass is interfering with the Session's internals in some way).   this is why the error is there.

> 
>     2. The "100" limit seems arbitrary. Is it ever ok for 88 flush()es to be necessary here? 

if custom event handling code is causing additional flushes to occur, this is not necessarily an error as there can be routines that add additional items to be flushed, however a reasonable number of additional flushes would usually be less than ten.

> If so, why not 101? Why not throw after the first one? In other words, is it absolutely incorrect behavior in a program to "create new objects" or whatever it is that is actually causing _is_clean() to return false?

it's only incorrect behavior if the code which is making this happen has no guards in place to keep from adding new items as a result of the previous operation without ever stopping.  100 is a number that suggests the code has no way to stop what it's doing.   if the code flushed 57 times and stopped, that would be very unusual, unless it was very specific.

You could make the same argument for Python's recursion overflow error.  It defaults to a stack depth of 1000.  Why not 1001?  

> 
>     3. I'm using a custom session object that has some facilities to help in gevent/multi-processing SQLA usage. Possibly that's what's interfering with the _is_clean(). Can someone give me a high level idea of what the code in _prepare_impl(), which is attempting 100 flushes is intended to do?

prepare_impl() is (very) loosely analogous to the "prepare" phase of a two-phase commit operation (2PC).    The idea is that when prepare() succeeds, all operations that the Session needs to do in order to commit the transaction have completed successfully, and it is now safe to commit.

one day many years ago, someone came to us with a custom event handler, probably a session.after_flush() handler, that looked at the state of objects and based on some logic would then decide that there were some additional database changes that needed to occur; it would then modify the state of the session further in some way (either adding new objects or modifying attributes on existing objects).   This then grows out the "new", "dirty" and "deleted" collections.     What would happen is when the user code would call session.commit(), the prepare step would happen first, the final flush() of objects would proceed (invoked only once), and then COMMIT would happen, and afterwards the session would reset all its internal state and consider itself clean (or maybe it raised an error, not sure), while in fact there were new, pending changes that had been produced within the prepare phase itself.

So the decision was made that we could add a check within prepare to see if event handlers did in fact produce more changes after flush was otherwise complete, and that we could call flush() again, to get those changes in as well.   The details of the discussion are likely here in a migrated issue from long ago that you could get from the git history.   I dont recall if the "stop at 100" guard was added at the same time but it likely was.   

> 
> EDITED: _is_clean() seems to return True is there are no modifications in the identity_map, no new and no deleted objects, I guess. 

that's waht it does.

> If the flush() action happening in the _prepare_impl() ends up being a NOOP in my subclass of Session that I suppose would lead to this behavior, because then the internal flush doesn't affect the status of any dirty/etc objects...

session subclass, seems like a likely factor in this error you're getting.

**Reply by aathan** [reply]
I haven't completely read through your answer, but since I have your attention I wanted to immediately respond to at least the part of your answer regarding item 1):

```
    def _prepare_impl(self):
...
	if not self.session._flushing:
            for _flush_guard in range(100):
		if self.session._is_clean():
                    break
		self.session.flush()
            else:
                raise exc.FlushError(
                    "Over 100 subsequent flushes have occurred within "
                    "session.commit() - is an after_flush() hook "
                    "creating new objects?"
```

```
   def commit(self, _to_root=False):
        self._assert_active(prepared_ok=True)
        if self._state is not PREPARED:
            self._prepare_impl()
```

The code in _prepare_impl() and commit() in Session.py would seem to directly contradict the notion that the existing error text isn't misleading. The session itself is the proximal cause of the 100 flush()es irrespective of whether they would be necessary due to activity in the client code. My suggestion would be to change this text to read something more like this, because (as you note and as I did) SQLA users may actually subclass Session.

"session.prepare() or session.commit() has now attempted 100 flush()es to bring the session into clean state - is an after_flush() hook creating new objects?"

The point here being to make it crystal clear that it is not user code that is calling flush().

EDIT: ... and in fact, even if user code was calling flush() in addition to the code in session itself, those are not counted toward the 100.

**Reply by zzzeek** [reply]
that's fine language to use, though I dont see where the existing message is mentioning anything about the end user calling flush()   It only refers to the after_flush() event hook, which is called within commit() as well as prepare().    The actual .prepare() method on Session is almost never used so mentioning it in the message should at least be parenthesized.

**Reply by zzzeek** [reply]
but yes I like that language "to bring the session into a clean state", that's great.

I originally had the impression that you disagreed with the suggestion that an end-user after_flush() hook is a likely culprit for this issue.

**Reply by aathan** [reply]
Totally. And thank you for your time and explanations. I agree it's not explicitly stated, all I'm saying is that the "Over 100 subsequent flushes have occurred within session.commit" when combined with the suggestion that it's a flush hook that's causing the problem, left me with the impression the flush()es in question could be in my code.

As it turns out, it seems likely my problem is that my flush() is a NOOP relative to the "try flushing 100 times" heuristic in SQLA's session preparation.

**Comment by aathan** [comment]
Regarding your answer to 2)  -- I understand now that the 100 flush limit is a reasonable heuristic to detect bad behavior. It just wasn't clear to me from the error message that this wasn't some enshrined limit on the actual number of flush() calls that can be made during a commit. Instead it is a behavior of the prepare()ing of sessions done either explicitly by calling prepare() or implicitly by commit() when the session isn't in the prepared state already. IMHO my suggested change to the error message makes that issue clear.

Regarding the rest of the items:  Yes ok, and thank you.

The proximal cause appears to be that the session is not in fact prepared() and that the flush()es done to prepare it don't progress its state, due to my override of prepare() etc... willfix (in my code).

The error message produced gave the impression the flush()es are occurring in my code, not SQLA's.

This may motivate a change to the error message.

**Comment by aathan** [comment]
I've submitted a PR with "improved" error text.

**Comment by aathan** [comment]
Not sure you'll see this, but looking into the problem further, I'm realizing that at some point I added an explicit call to flush() in my subclass' commit() -- it had previously not been needed, and the comment I left for myself says it was needed for SQLA 1.4 .  The reason (SQLA 1.4) may be a red-herring. Really it may have been some other state-dependent change not directly related to the SQLA version.

Anyway, looking at SessionTransaction.commit() it would appear the explicit flush() would have been necessitated if the Session was already PREPARED because this would then have caused _prepare_impl() to skip the flush(). Any hints on major changes that may have affected the PREPARED state of a Session?

Also, in looking at the code I see the `no_autoflush()` context manager.  It's clearly documented as disabling "flushes occuring upon query access" but the name no_autoflush() seems broader. From what I can tell only merge() (and some associated code) uses Session._autoflush(). (I made the addition to the PR I submitted recently, in anticipation of "yes")

Would it therefore be correct to add a note that "no_autoflush() will not disable the automatic flush() that occurs during a commit()." ? This may seem redundant to an SQLA expert, but may be a useful note for new users.

**Reply by zzzeek** [reply]
we can add more of a note at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#flushing.  The docs state: "The “flush-on-Query” aspect of the behavior can be disabled by constructing [sessionmaker](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.sessionmaker) with the flag autoflush=False:".    But certainly, we can add a note that the term "autoflush" does not refer to the "flush" that occurs within .commit(), which is always necessary.

as for the rest of your comment, it's hard to follow without knowing what it is you're trying to do.

**Comment by aathan** [comment]
I think we're all set on this discussion. Thanks again for the help and input.

**Reply by zzzeek** [reply]
thanks for being a good sport about it (edit: i mean that in a generous way.   I've been pretty opinionated here).
